R/Quosure.R

```diff
@@ -0,0 +1,270 @@
+#' Reproducible class with environment and code.
+#'
+#' @slot code (`character`) to reproduce the environment
+#' @slot env (`environment`) environment which content was generated by the evaluation
+#' of the `code` slot.
+setClass(
+  "Quosure",
+  representation(code = "character", env = "environment"),
+  prototype(code = character(0), env = new.env(parent = parent.env(.GlobalEnv)))
```

Comment:
I would suggest to use expression / call type instead which is much more closer to the _code_ representation in R.

Reply:
We should rather handle code as character as it supports comments, empty lines etc. `language` classes discards comments information.

Reply:
The question is: do we really need to have a comments, empty lines etc? The value is very questionable for me. I feel that it is a _code_ that we need to capture, evaluate and return - not a _script_ that produces is. I also think that it would be better to stick to the more raw and simple structures as managing characters might be unpredictable and could lead into some unforeseen issues such as char encoding, empty lines etc. Character should be _just_ a representation of code at the very end and such casting should be done by shiny module - not the class.

Reply:
Are you aware that current implementation which returns SRC with comments and new-line-spaces will not be possible anymore?
What in your opinion would be unpredictable?:
- `"var <- 1"` - code as character
- `"# this is a comment"` - code (comment) as character
- `""` - empty line - code (empty) as character

We will support primarily:
- `var <- 1` - unquoted expression (to be substituted)
- `substitute(var <- 1, list(var = quote(new_var))` - quoted expression
- `rlang::expr(!!new_var <- 1)` - quoted expression

To possibly add a comment one would have to use `parse` which will return an empty `expression()` but we can get parsed text from the attr(expression, "wholeSrcref")
```r
parse(text = "
# comment

# new comment
")

#> expression()
```



Accepting only expression doesn't prevent anyone from executing `"\n"` which is a valid R expression and will add an artifact to the code (it won't be extra line but "\n" visible as is in the code.
```r
eval_code(Quosure, quote("\n"))
```

Reply:
> Are you aware that current implementation which returns SRC with comments and new-line-spaces will not be possible anymore?

Yes I am and this is what I put in question - I am not so sure if it brings a lot of value to the users but it changes drastically the implementation. The problems with char management differs significantly from problems of code management and I feel that code management would be easier, simpler and closer to the actual task.
An example of unexpected issues is regular expressions applied on code as char which doesn't seem to be an appropriate way of managing the code. If we store code as char then what is stopping devs to manipulate the data with regular expression? Well... nothing! Same for creating inputs for this class. I want module developers to create code/expressions - not chars. Expressions are safer and char can accept anything. See below:
```
> expression(1a1)
Error: unexpected symbol in "expression(1a1"
> "1a1"
[1] "1a1"
```

> To possibly add a comment one would have to use ...

How about no comments at all? At the end of the day, SRC returns a _code_ not a _script_. I know I representing some extreme view right now (and I am feeling like a devil advocate) but I need to ask some basic questions because we might just blindly assume something that is not true (anymore).

Reply:
There is some way to extract the `srcref` from the parsed expression to have the comments or the line-breaks so it can be preserved (and combined) in the expr attribute. This is diffucult task but possible In my opinion to achieve. Only adding comments and new line would have to be done through `eval_code(parse(text = "any code"))`.

```r
parse(text = "
  #

  elo <- 1
") |> attr("wholeSrcref")

#> #
#>
#> elo <- 1
```